tools/scripts/check_links.py

```diff
@@ -0,0 +1,238 @@
+    html_links = [a.get('href') for a in soup.find_all('a', href=True)]
+
+    # Extract src attributes from <img> tags
+    html_images = [img.get('src') for img in soup.find_all('img', src=True)]
+
+    # Combine all links
+    all_links = links + images + html_links + html_images
+
+    # Filter out None values and duplicates
+    all_links = list(set(filter(None, all_links)))
+
+    return all_links
+
+def is_external_link(url):
+    """
+    Check if a URL is an external link.
+    """
+    parsed_url = urlparse(url)
+    return parsed_url.scheme in ('http', 'https', 'ftp')
+
+def get_md_files_to_links(md_files, include_external):
+    """
+    Create a dictionary mapping Markdown files to their extracted links.
+    """
+    md_files_to_links = {}
+    for md_file in md_files:
+        links_in_md_file = extract_links_from_markdown(md_file)
+        links_to_inspect = []
+        md_dir = os.path.dirname(md_file)
+
+        for link in links_in_md_file:
+            link = link.strip()
+
+            if link in bypass_inspection_links:
+                continue
+            if not link or link.startswith('#'):
+                continue
+            if link.startswith(('mailto:', 'tel:')):
+                continue
+            if is_external_link(link) and not include_external:
+                continue
+
+            links_to_inspect.append(link)
+
+        links_to_inspect = list(set(links_to_inspect))
+        md_files_to_links[md_file] = links_to_inspect
+
+    return md_files_to_links
+
+def compute_abs_link(md_dir, link):
+    """
+    Compute the absolute path of a link, handling anchors and normalizing the path.
+    """
+    link_path = link.split('#')[0]
+    abs_path = os.path.normpath(os.path.join(md_dir, link_path))
+    abs_path = abs_path.lstrip("/\\")
+    return abs_path
+
+def check_link(session, link, md_file, headers, link_inspection_results):
+    """
+    Inspect a single link, returning the result.
+    """
+    md_dir = os.path.dirname(md_file)
+    result = None
+    abs_link = link
+
+    # Cache key includes md_dir for internal links to avoid conflicts
+    cache_key = link if is_external_link(link) else os.path.normpath(os.path.join(md_dir, link))
+
+    # Check if the link has already been inspected
+    if cache_key in link_inspection_results:
+        result = link_inspection_results[cache_key]
+        if not is_external_link(link):
+            abs_link = compute_abs_link(md_dir, link)
+    else:
+        if is_external_link(link):
+            try:
+                response = session.head(link, allow_redirects=True, timeout=10, headers=headers)
+                if response.status_code != 403 and (response.status_code >= 400 or response.status_code == 405):
+                    response = session.get(link, allow_redirects=True, timeout=10, headers=headers)
+                    if response.status_code >= 400 and response.status_code != 403:
+                        result = response.status_code
```

Comment:
This needs more documentation to explain why `403` is excluded from consideration.

Reply:
This has been updated to include documentation and to be clearer.